File: src/javascript/algorithms/strings/4.validAnagram/validAnagram.js
/**
 * Checks if two strings are anagrams using a frequency array (lowercase a-z only).
 * TC: O(n), SC: O(1)
 * @param {string} str1
 * @param {string} str2
 * @returns {boolean}
 */
function validAnagram1(str1, str2) {
    if(str1.length !== str2.length) return false;

    let charFrequencyArr = new Array(26).fill(0);

    for(let i=0; i< str1.length; i++) {
        let charCode1 = str1.charCodeAt(i);
        let charCode2 = str2.charCodeAt(i);
        charFrequencyArr[charCode1-97]++;
        charFrequencyArr[charCode2-97]--;
    }

    return charFrequencyArr.every((n) => n === 0);
}

/**
 * Checks if two strings are anagrams using a Map (works for any characters).
 * TC: O(n), SC: O(n)
 * @param {string} str1
 * @param {string} str2
 * @returns {boolean}
 */
function validAnagram2(str1, str2) {
    if(str1.length != str2.length) return false;

    let charMap = new Map();

    for(let ch of str1) {
        charMap.set(ch, (charMap.get(ch) || 0) + 1);
    }

    for(let ch of str2) {
        if(charMap.has(ch)) {
            charMap.set(ch, charMap.get(ch)-1);
            if(charMap.get(ch) === 0) charMap.delete(ch);
        } else {
            return false;
        }
    }
     
    return charMap.size === 0;
}

/**
 * Checks if two strings are anagrams by sorting and comparing.
 * TC: O(n log n), SC: O(n)
 * @param {string} str1
 * @param {string} str2
 * @returns {boolean}
 */
function validAnagram3(str1, str2) {
    if(str1.length !== str2.length) return false;

    return [...str1].sort().join('') === [...str2].sort().join('');
}



// Test cases
const testCases = [
    { str1: "anagram", str2: "nagaram", expected: true },
    { str1: "rat", str2: "cat", expected: false },
    { str1: "Listen", str2: "Silent", expected: true },
    { str1: "aabbcc", str2: "abcabc", expected: true },
    { str1: "aabbcc", str2: "aabbc", expected: false },
    { str1: "123", str2: "321", expected: true },
    { str1: "a!b@c", str2: "c!b@a", expected: true },
];

for (const { str1, str2, expected } of testCases) {
    console.log(
        `Input: "${str1}", "${str2}" | Array: ${validAnagram1(str1, str2)} | Map: ${validAnagram2(str1, str2)} | Sort: ${validAnagram3(str1, str2)} | Expected: ${expected}`
    );
}

File: src/python/algorithms/strings/4.validAnagram/valid_anagram.py
"""
Valid Anagram

Given two strings s and t, return true if t is an anagram of s.

Time Complexity: O(n)
Space Complexity: O(1) - fixed 26 characters
"""

from collections import Counter


def is_anagram(s, t):
    """
    Using Counter.
    TC: O(n), SC: O(1)
    """
    return Counter(s) == Counter(t)


def is_anagram_array(s, t):
    """
    Using character count array.
    TC: O(n), SC: O(1)
    """
    if len(s) != len(t):
        return False
    
    count = [0] * 26
    
    for i in range(len(s)):
        count[ord(s[i]) - ord('a')] += 1
        count[ord(t[i]) - ord('a')] -= 1
    
    return all(c == 0 for c in count)


def is_anagram_sort(s, t):
    """
    Using sorting.
    TC: O(n log n), SC: O(n)
    """
    return sorted(s) == sorted(t)


# Test cases
if __name__ == "__main__":
    test_cases = [
        {"s": "anagram", "t": "nagaram"},  # True
        {"s": "rat", "t": "car"},           # False
    ]
    
    for test in test_cases:
        print(f"s: '{test['s']}', t: '{test['t']}'")
        print(f"Is anagram: {is_anagram(test['s'], test['t'])}")
        print()
